// nodame/core/index.js
/**
 * @author  Argi Karunia <arugikaru@yahoo.co.jp>
 * @link    https://github.com/tokopedia/nodame
 * @license http://opensource.org/licenses/MIT
 *
 * @version 1.0.0
 */

var numeral = require('numeral');
var path    = require('path');

module.exports = (function () {
    var settings        = {};
    var __argv          = require('nodame/argv');
    var __env           = require('nodame/env');

    var sysPath = function () {
        return path.normalize(__dirname + '/../../../..');
    };

    var appPath = function () {
        return path.normalize(sysPath() + '/../..');
    };

    var init = function () {
        settings.__systems  = {};

        // Init argv
        settings.__systems.argv = __argv.set();

        // Init config
        var argvCopy        = settings.__systems.argv;
        var cfg             = require('nodame/config');
        var cfgs            = cfg.parse(argvCopy, appPath());
        settings.config     = cfgs.configs;
        settings.__systems.configPath = cfgs.path;

        // console.log(settings.config);
        // throw new Error();

        // Init environment
        var releaseEnv      = settings.config.server.release_env;
        settings.__systems.env  = __env.set(settings.__systems, releaseEnv);
    };

    init();

    var argv = settings.__systems.argv;

    var set = function (key, obj) {
        if (key === undefined || obj === undefined) {
            throw new Errors('Invalid settings set parameters.');
        }

        settings[key] = obj;
    };

    var express = require('express');

    var router  = function () {
        return this.express.Router();
    };

    var config  = function (key) {
        var read = function (obj, params) {
            if (params.length === 0) {
                return obj;
            }

            if (params.length > 1) {
                obj = obj[params[0]];
                params.shift();
                return read(obj, params);
            } else {
                obj = obj[params[0]];
                return obj;
            }
        };

        var __config = function () {
            key     = key.split('.');
            var val = read(settings.config, key);

            return val;
        };

        return __config();
    };

    var __getFilePath = function (module, name) {
        return path.normalize(sprintf('%s/%s/%s', appPath(), module, name));
    };

    /**
     *  Native's require wrapper
     *
     *  This method is required to call nodame's modules and custom modules
     *
     *  @param {string} name Module's name or path
     *  @return {object} module
     */
    var __require = function (name) {
        var vars = name.split('/');

        if (vars[0] === 'my_modules') {
            return require(__getFilePath('modules', name));
        } else {
            return require(name);
        }
    };

    /**
     *  Service loader
     *
     *  Load application's service
     *
     *  @param {string} name Module's name or path
     *  @return {object} module
     */
    var service = function (name) {
        return require(__getFilePath('services', name));
    };

    /**
     *  Handler loader
     *
     *  Load application's handler
     *
     *  @param {string} name Module's name or path
     *  @return {object} module
     */
    var handler = function (name) {
        return require(__getFilePath('handlers', name));
    };

    /**
     *  Middleware loader
     *
     *  Load application's middleware
     *
     *  @param {string} name Module's name or path
     *  @return {object} module
     */
    var middleware = function (name) {
        return require(__getFilePath('middlewares', name));
    };

    /**
     *  Enforce mobile views
     *
     *  Load mobile's view. This is a middleware
     *
     *  @return {object} middleware
     */
    var enforceMobile = function () {
        var self = this;

        return function (req, res, next) {
            switch (ENFORCE_MOBILE) {
                case 'soft':
                    req.device.type = 'phone';
                    break;
                case 'hard':
                    req.device.type = 'desktop';
                    html = self.require('nodame/html').new(req, res);
                    html.headTitle(settings.config.app.title);
                    html.headDescription(settings.config.app.desc);
                    html.render({
                        module: 'errors',
                        file: 'interrupt'
                    });
                    break;
            }

            if (next && ENFORCE_MOBILE !== 'hard') return next();
        };
    };

    /**
     *  View's variables setter
     *
     *  Set variable to view's locals. This is a middleware.
     *
     *  @param {obj} app Express
     *  @return {object} middleware
     */
    var locals = function (app) {
        var self = this;
        var qs = require('query-string');
        var configs = {
            url: this.config('url'),
            app: this.config('app'),
            assets: this.config('assets')
        };
        var assets = this.settings.assets;

        return function (req, res, next) {
            app.locals.nodame = {
                time: function (num) {
                    return String(num).replace(/([0-9]{2})([0-9]{2})/, '$1:$2 WIB');
                },
                /**
                 *  Lead zero
                 *
                 *  Convert number to use leading zero format
                 *
                 *  @param {int} num Number
                 *  @return {string} number
                 */
                leadZero: function (num) {
                    return ('0' + num).slice(-2);
                },
                date: function (date, format) {
                    format = format || 'normal';
                    date = new Date(date);
                    var weekdays = {
                        Sun: 'Minggu',
                        Mon: 'Senin',
                        Tue: 'Selasa',
                        Wed: 'Rabu',
                        Thu: 'Kamis',
                        Fri: 'Jumat',
                        Sat: 'Sabtu'
                    };

                    var months = {
                        Jan: {
                            word: 'Januari',
                            num: '1'
                        },
                        Feb: {
                            word: 'Februari',
                            num: '2'
                        },
                        Mar: {
                            word: 'Maret',
                            num: '3'
                        },
                        Apr: {
                            word: 'April',
                            num: '4'
                        },
                        May: {
                            word: 'Mei',
                            num: '5'
                        },
                        Jun: {
                            word: 'Juni',
                            num: '6'
                        },
                        Jul: {
                            word: 'Juli',
                            num: '7'
                        },
                        Aug: {
                            word: 'Agustus',
                            num: '8'
                        },
                        Sep: {
                            word: 'September',
                            num: '9'
                        },
                        Oct: {
                            word: 'Oktober',
                            num: '10'
                        },
                        Nov: {
                            word: 'November',
                            num: '11'
                        },
                        Dec: {
                            word: 'Desember',
                            num: '12'
                        }
                    };

                    return date.toString().replace(/^([a-z]{1,3}) ([a-z]{1,3}) ([0-9]{1,2}) ([0-9]{1,4}) ([0-9]{1,2}):([0-9]{1,2}):([0-9]{1,2}) GMT([-+][0-9]{1,4}) (\([a-z ]+\))$/i, function (match, $1, $2, $3, $4, offset, original) {

                        var res = '';

                        switch (format) {
                            case 'normal':
                                res = weekdays[$1] + ', ' + $3 + ' ' + months[$2].word + ' ' + $4;
                                break;
                            case 'short':
                                res = ('0' + $3).slice(-2) + '/' + ('0' + months[$2].num).slice(-2) + '/' + $4;
                                break;
                        }

                        return res;
                    });
                },
                thousands: function (num) {
                    return numeral(num).format(0.0);
                },
                locale: function (data) {
                    var html = '';

                    if (data !== undefined && Array.isArray(data) && data.length > 0) {
                        var locales = {};

                        for (var i in data) {
                            var key = data[i];
                            locales[key] = res.locals.__(key);
                        }

                        localesData = 'var b=' + JSON.stringify(locales) + ';';

                        html += '<script type="text/javascript">';

                        html += 'function __(a) {' + localesData + 'return b[a];}';
                        html += '</script>';
                    }

                    return html;
                },
                url: {
                    base: function (uri, params) {
                        uri = path.normalize('/' + uri);
                        var url = configs.url.base + uri;
                        return this.stringify(url, params);
                    },
                    assets: function (uri, params) {
                        uri = path.normalize('/' + uri);
                        var url = configs.url.assets + uri;
                        return this.stringify(url, params);
                    },
                    stringify: function (url, params) {
                        return url + this.__getQueries(params);
                    },
                    __getQueries: function (params) {
                        var queries = '';

                        if (Object.keys(params).length > 0) {
                            queries = '?' + qs.stringify(params);
                        }

                        return queries;
                    }
                },
                assets: {
                    __device: 'mobile',
                    __type: 'css',
                    __module: undefined,
                    css: function (module) {
                        return this.__assets('css', module);
                    },
                    js: function (module) {
                        return this.__assets('js', module);
                    },
                    __assets: function (type, module) {
                        var device = 'desktop';

                        if (req.device.type !== 'desktop') {
                            device = 'mobile';
                        }

                        this.__device = device;
                        this.__type = type;
                        this.__module = module;
                        var assetsName = this.__getValidName();
                        var assets = this.__html(assetsName);
                        return assets;
                    },
                    __getValidName: function () {
                        var type = this.__type;
                        var name = this.__getName(this.__module || 'global');

                        name = name.replace(/[.][a-z]+$/gi, '');
                        var re = new RegExp(name, 'gi');

                        for (var i in assets[type]) {
                            for (var __name in assets[type][i]) {
                                if (__name.match(re)) {
                                    return __name;
                                }
                            }
                        }

                        return this.__getName('global');
                    },
                    __getConfigDir: function () {
                        var confDir = configs.assets.dir;

                        if (confDir.substr(0, 1) == '/') {
                            confDir = confDir;
                        } else {
                            confDir = path.normalize(self.appPath() + '/' + confDir);
                        }

                        return confDir;
                    },
                    __getName: function (module) {
                        var confDir = this.__getConfigDir();

                        return path.normalize(confDir + '/min/' + this.__device + '.' + module + '.min.' + this.__type);
                    },
                    __html: function (name) {
                        var type = this.__type;
                        var data = assets[type];
                        var html = '';

                        if (self.isDev()) {
                            for (var i in data) {
                                if (data[i][name] !== undefined) {
                                    if (self.isDev()) {
                                        for (var j in data[i][name]) {
                                            var filepath = data[i][name][j];
                                            html += this.__htmlTag(type, filepath);
                                        }
                                    } else {
                                        html = this.__htmlTag(type, name);
                                    }
                                }
                            }
                        } else {
                            html = this.__htmlTag(type, name);
                        }

                        return html;
                    },
                    __htmlTag: function (type, filepath) {
                        var html = '';
                        var configDir = this.__getConfigDir();
                        filepath = filepath.replace(configDir, configs.url.assets);

                        switch (type) {
                            case 'css':
                                html = '<link href="' + filepath + '" type="text/css" rel="stylesheet">\n';
                                break;
                            case 'js':
                                html = '<script src="' + filepath + '" type="text/javascript"></script>\n';
                                break;
                        }

                        return html;
                    }
                }
            };

            if (next) return next();
        };
    };

    return {
        appPath         : appPath,
        argv            : argv,
        config          : config,
        enforceMobile   : enforceMobile,
        env             : __env.env,
        express         : express,
        handler         : handler,
        isDev           : __env.isDev,
        locals          : locals,
        middleware      : middleware,
        require         : __require,
        router          : router,
        service         : service,
        set             : set,
        settings        : settings,
        sysPath         : sysPath
    };
})();
